webrtc/browser-to-browser.md

```diff
@@ -0,0 +1,122 @@
+
+7. Messages on `RTCDataChannel`s on the established `RTCPeerConnection` are framed using the message framing mechanism described in [multiplexing].
+
+## STUN
+
+A node needs to discover its public IP and port, which is forwarded to the remote node in order to connect to the local node.
+On non-browser libp2p nodes doing a hole punch with TCP or QUIC, the libp2p node discovers its public address via the [identify] protocol.
+One can not use the [identify] protocol on browser nodes to discover ones public IP and port given that the browser uses a new port for each connection.
+For example say that the local browser node establishes a WebRTC connection C1 via browser-to-server to a server node and runs the [identify] protocol.
+The returned observed public port P1 will most likely (depending on the NAT) be a different port than the port observed on another connection C2.
+The only browser supported mechanism to discover ones public IP and port for a given connection is the non-libp2p protocol STUN.
+This is why this specification depends on STUN, and thus the availability of one or more STUN servers for _A_ and _B_ to discovery their public addresses.
+
+There are various publicly available STUN servers.
+As an alternative one may operate dedicated STUN servers for a given libp2p network.
+Further specification of the usage of STUN is out of scope for this specifitcation.
+
+As an aside, note that _A_ and _B_ do not need to use the same STUN server in order to establish a direct WebRTC connection.
+
+## Signaling Protocol
+
+The protocol id is `/webrtc-direct`.
+Messages are sent prefixed with the message length in bytes, encoded as an unsigned variable length integer as defined by the [multiformats unsigned-varint spec][uvarint-spec].
+
+``` protobuf
+syntax = "proto2";
+
+message Message {
+    // Specifies type in `data` field.
+    enum Type {
+        // String of `RTCSessionDescription.sdp`
+        OFFER = 0;
+        // String of `RTCSessionDescription.sdp`
+        ANSWER = 1;
+        // String of `RTCIceCandidate.toJSON()`
+        CANDIDATE = 2;
+    }
+
+    optional Type type = 1;
+    optional string data = 2;
+}
+```
+
+## Open Questions
+
+- Do we need a mechanism for browsers to advertise support for WebRTC browser-to-browser?
+
+  Say that browser B supports WebRTC browser-to-browser.
+  B listens via a relay and advertises its relayed address.
+  A discovers B's relayed address.
+  At this point A does not know whether B is a browser and thus supports WebRTC browser-to-browser, or whether B is e.g. a laptop potentially supporting TCP and QUIC hole punching via DCUtR but not WebRTC browser-to-browser.
+  In the latter case, A can not establish a direct connection to B.
+
+  Potential solution would be for B to advertise some protocol after the `/p2p-circuit` within its Multiaddr, e.g. `/ip6/<RELAY_IP>/udp/4001/p2p/<RELAY_PEER_ID>/p2p-circuit/webrtc-direct/p2p/<B_PEER_ID>`.
```

Comment:
Something of the form `/ip6/<RELAY_IP>/udp/4001/p2p/<RELAY_PEER_ID>/p2p-circuit/webrtc-direct/p2p/<B_PEER_ID>` is a bit weird, but I think it's consistent with [how multiaddrs are interpreted](https://github.com/multiformats/multiaddr#interpreting-multiaddrs). The `p2p` component is passed into `webrtc-direct` which then uses the rest of the multiaddr (to the left) to make the connection.

Does this mean that if a node is listening on N relays, it will have 2*N multiaddrs to share?

Maybe part of the weirdness is that we want to tell others we support the webrtc-direct transport, but that transport is always started by piggy backing on some other transport. What if we had _B_ advertise `/webrtc-direct/p2p/<B_PEER_ID>` along with their other multiaddrs? Then A knows _B_ speaks `webrtc-direct` and knows it can use any other known way to reach B to set that up.

Reply:
> Maybe part of the weirdness is that we want to tell others we support the webrtc-direct transport, but that transport is always started by piggy backing on some other transport. What if we had _B_ advertise `/webrtc-direct/p2p/<B_PEER_ID>` along with their other multiaddrs? Then A knows _B_ speaks `webrtc-direct` and knows it can use any other known way to reach B to set that up.

It is an interesting idea. To paraphrase: Our signaling protocol doesn't care about the underlying transport. It just happens that the whole point of `/webrtc-direct` is to establish a direct connection **because** we don't have any other means so very likely, it will always be paired with listening only on a relay address.

A funny example to make a point: If we had message-based communication (in addition to streams), we could imagine a multiaddr in the form of: `/email/foo@bar.com` and run the signaling protocol over that one.

On the other hand, publishing `/webrtc-direct/p2p/<PEER_ID>` as a listening address feels weird because you cannot dial that one by itself.

Really, isn't it what we want to advertise that we support `/webrtc` as a transport technology and in combination with supporting the signaling protocol, it means we can perform the necessary communication to attempt a direct connection?

Reply:
This specification already requires the use of STUN. Would it make sense to use STUN to figure out IP and port(s) and advertise itself via `/webrtc` just without the `/certhash`?

Technically, that is what we connect to. It just requires further out-of-band knowledge which is why we have the signaling protocol.

Reply:
We also trickle in candidates/info, right? And that information may become stale (? Not sure about this)

> On the other hand, publishing /webrtc-direct/p2p/<PEER_ID> as a listening address feels weird because you cannot dial that one by itself.

The node is listening on that protocol. It just happens to use some other mechanism besides a direct tcp/udp connection. This kind of highlights the flexibility of multiaddrs. We don’t have to assume this connection will happen directly over a tcp/udp conn.

Reply:
> We also trickle in candidates/info, right? And that information may become stale (? Not sure about this)

Yes, we'd have to adapt the signaling protocol if we go down that path.

> > On the other hand, publishing /webrtc-direct/p2p/<PEER_ID> as a listening address feels weird because you cannot dial that one by itself.
> 
> The node is listening on that protocol. It just happens to use some other mechanism besides a direct tcp/udp connection. This kind of highlights the flexibility of multiaddrs. We don’t have to assume this connection will happen directly over a tcp/udp conn. 

Yes-ish 😁
Wasn't one of the other design aspects of multiaddresses that they are self-descriptive and contain all the information necessary to establish a connection, i.e. no out-of-band knowledge?

Reply:
> This specification already requires the use of STUN. Would it make sense to use STUN to figure out IP and port(s) and advertise itself via `/webrtc` just without the `/certhash`?
> 
> Technically, that is what we connect to. It just requires further out-of-band knowledge which is why we have the signaling protocol.

This won't be possible for the following reasons:
1. The IP address of the UDP socket for the host candidate (the candidate on a local interface) cannot be specified in a browser.
1. Candidate gathering only starts after SDP exchange is complete. The initial SDP in most cases will not contain any candidates, or will only contain host candidates which are not useful for connecting from an external network.
1. If we figure out a public IP using STUN and publish that as a multiaddr, browsers cannot multiplex WebRTC over the single UDP socket the same way as Pion.

Reply:
Thanks for the input here!

To summarize, we seem to have two options thus far:

1. Advertise `ip6/<RELAY_IP>/udp/4001/p2p/<RELAY_PEER_ID>/p2p-circuit/webrtc-direct/p2p/<B_PEER_ID>`
    - In case the advertising node supports any other protocol besides the WebRTC signaling protocol on top of a relayed connection, it needs to advertise X*N  Multiaddresses, with X being the number of protocols and N being the number of relays.
2. Advertise `/webrtc-direct/p2p/<PEER_ID>`
    - Not in itself dialable.

> This specification already requires the use of STUN. Would it make sense to use STUN to figure out IP and port(s) and advertise itself via `/webrtc` just without the `/certhash`?

Not listing this idea above. The IP is undialable, given that the node is likely behind a NAT and/or firewall. The port, if on a browser node, changes per connection and thus outdated at the point of advertisement.

Reply:
On one meta level above, I don't think an answer to this open question is strictly necessary in order for this specification to be useful. This specification in combination with the trial-and-error approach via a relayed connection + identify or an optimistic `/webrtc-direct` stream, while less efficient, can already be used on a network.

With the above in mind, I suggest not blocking this specification on resolving this answer. That said, let's keep discussing here while the pull request is still open.

Reply:
@mxinden Does X*N take into account DCUtR? Because if yes, for browsers this will be limited to 2*N connections as they will only support relayed connections and webrtc.
AFAICT, publishing addresses does not seem frequent enough where 2*N should be too much of a problem.

Reply:
Circuit addresses seem valid without a p2p multiaddress following `/p2p-circuit` as referenced [here](https://github.com/multiformats/js-mafmt/blob/master/src/index.ts#L108), so adding an additional protocol should be easy.

Reply:
> laptop potentially supporting TCP and QUIC hole punching via DCUtR but not WebRTC browser-to-browser.

Long-term, I expect/hope that a laptop will also be able to use this protocol to connect to/from browsers (which again, highlight that "browser-to-browser" isn't a good name)

So no matter if it's a laptop or a browser, you would connect to it using a relay, and then negotiate `dctur` or `webrtc` using multistream select.

Sure, for a while you may stumble onto nodes that only support `dctur` and that will fail, but should we optimize for that case?

Reply:
> In practice, I don't expect this to be a big problem, since nodes will typically only use one or two relays.

Let's say I'm behind a NAT on a desktop application. I use two relays, each of them have every transports available, I need to advertise:
(relay transports: tcp + wss + webrtc + webtransport + quic) * (numrelays: 2) * (holepunch mechanisms: dctur + webrtc) = 20 multi addresses.

Each of those will be at least 100 bytes, so 2kb total

> It's a shame that we don't do this for TCP and QUIC!

It would become:
(relay transports: tcp + wss + webrtc + webtransport + quic) * (numrelays: 2) * (target peer transports: quic + tcp + websocket) = 30 multi addresses.

3kb isn't _that_ bad, but it still feels wrong. And we'll be able to multiply that by two again when we have ipv6 widely deployed.

Let's compare to something else, let's say we add two fields to the PeerRecord:
```protobuf
message PeerRecord {
  message AddressInfo {
    bytes multiaddr = 1;
  }

  bytes peer_id = 1;
  uint64 seq = 2;

  repeated AddressInfo addresses = 3;
  repeated PeerRecord relays = 4; // new field
  repeated string capabilities = 5; // new field
}
```

I'll put here a record example in json for readability:
```json
{
  "peer_id": "QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7DWjhx5N",
  "relays": [
    {
      "peer_id": "Qmc3PxhMhQja8N4t7mRDyGm2vHkvcxe5Kabp2iAig1DXHb",
      "addresses": [
        "/ip4/140.82.121.4/tcp/4440",
        "/ip4/140.82.121.4/tcp/4441/wss",
        "/ip4/140.82.121.4/udp/4442/webrtc/certhash/xxxxxx",
        "/ip4/140.82.121.4/udp/4443/quic",
        "/ip4/140.82.121.4/udp/4444/quic/webtransport",
      ]
    },
    {
      "peer_id": "Qmcfz2MaPjw44RfVpHKFgXwhW3uFBRBxByVEkgPhefKCJW",
      "addresses": [
        "/ip4/74.6.136.150/tcp/4440",
        "/ip4/74.6.136.150/tcp/4441/wss",
        "/ip4/74.6.136.150/udp/4442/webrtc/certhash/xxxxxx",
        "/ip4/74.6.136.150/udp/4443/quic",
        "/ip4/74.6.136.150/udp/4444/quic/webtransport",
      ]
    }
  ],
  "capabilities": ["tcp", "quic", "webrtc"]
}
```
Even in json, this is only 826 bytes, and will scale a lot better.

`capabilities` is the simplest form of this field, but ideally, we could make something even better, to allow for instance to distinguish which networks is a part of

> Not sure I subscribe to the argument „A multiaddr contains everything you need to connect to a node“.

I agree, and I don't even see that much value in that. Sure, a list of bootstrap nodes looks better with [multiaddresses](https://github.com/ipfs/kubo/blob/b8412efb80f1205be70dd71cfca64c68456682e8/config/bootstrap_peers.go#L18-L23) than with for instance [ENRs](https://github.com/eth-clients/eth2-networks/blob/a072c89b40cf4f28a26d047bb3aeb08995955f07/shared/mainnet/bootstrap_nodes.txt#L8-L28) from devp2p, because the values are human readable, but that's about it.

Reply:
@Menduist Coming at this from an implementation perspective, all current libp2p implementations consider the multiaddress to pick a transport to use for dialing. While it would be more compact to include the webrtc capability in the PeerRecord, implementation would be much simpler if support for webrtc signaling is advertised in a multiaddr.

Reply:
I just pushed c09392b documenting advertisement by appending a `/webrtc-direct` protocol component to ones relayed address, e.g. `/ip6/fe80::883:a581:fff1:833/udp/4001/quic/p2p/<relay-peer-id>/p2p-circuit/webrtc-direct/p2p/<b-peer-id>`.

@ckousik @MarcoPolo note that this is not exactly what we discussed in today's call as it adds the `/webrtc-direct` protocol component to the relayed address, instead of a separate multiaddr. Rational for me is that a single address (relay + `/webrtc-direct`) combines all information in a single address and thus does not require implementations to combine information of multiple addresses. I don't have a strong opinion on this, please comment here in case you do.

@Menduist I see your point of an advanced format like `PeerRecord` over the usage of multiaddr. I am not opposed to it, though I would like to keep these efforts (introducing webrtc browser-to-browser and a new address format) separate. For me the webrtc browser-to-browser effort has higher priority, thus pushing for extending multiaddr, at least for now. Let me know if that reasoning makes sense to you.

Reply:
> @ckousik @MarcoPolo note that this is not exactly what we discussed in today's call as it adds the /webrtc-direct protocol component to the relayed address, instead of a separate multiaddr. Rational for me is that a single address (relay + /webrtc-direct) combines all information in a single address and thus does not require implementations to combine information of multiple addresses. 

Not sure what was discussed in the call, but @mxinden's rationale makes sense to me. It shouldn't be necessary to combine multiple multiaddresses. In fact, that would break a lot of assumptions.

Can we please find a better name than `webrtc-direct` though? It's not direct, it's literally using a relay!

Reply:
I would like to avoid the name browser if possible.

Instead, what seems to be characterizing for the clients that benefit from this spec is that they don't control the WebRTC stack but have to talk to it through a standardized interface: `RTCPeerConnection`.

One can easily imagine that this interface is not just implemented by the browser but another "host" that can run "apps": WASM and wasmtime come to my mind here.

Most requirements in the spec are guided by limitations presented by this standardized interface. Thus I think it would be appropriate to tie the name to  this interface and call this "transport": `/webrtc-w3c`.

Reply:
Great idea. I suggest calling the multiaddr protocol component `/webrtc-w3c` and the signaling protocol multistream select identifier `/webrtc-w3c-signaling`.

Reply:
0a0a30a adopts the `webrtc-w3c` naming. Thanks @thomaseizinger.

I interpret the stalled advertisement discussion as silent agreement. Please comment in case you want to continue discussing it. Resolving this conversation. Please start a new thread in case you have more input.

//CC @MarcoPolo who advocated for multiple multiaddresses before.